module/zfs/zil.c

```diff
@@ -3195,6 +3195,18 @@ zil_commit(zilog_t *zilog, uint64_t foid)
 		return;
 	}
 
+	/*
+	 * The ->zl_commit_lock rwlock ensures that all in-flight
+	 * zil_commit() operations finish before suspension begins and that
+	 * no more begin. Without it, it is possible for the scheduler to
+	 * preempt us right after the zilog->zl_suspend suspend check, run
+	 * another thread that runs zil_suspend() and after the other thread
+	 * has finished its call to zil_commit_impl(), resume this thread
+	 * while zil is suspended. This can trigger an assertion failure in
+	 * VERIFY(list_is_empty(&lwb->lwb_itxs)).
+	 */
+	rw_enter(&zilog->zl_commit_lock, RW_READER);
```

Comment:
I also think we could use rw_tryenter() here.

Reply:
We need to take the lock to ensure that it is safe to read `zilog->zl_suspend`. How would rw_tryenter() help? We would still execute rw_enter() if it fails.

Reply:
We should fail to get reader lock only if there is a writer holding it (or trying to get it, depending on implementation).  And the writer lock is taken only when zl_suspend is going to be increased.  So I think we may fall back directly to txg_wait_synced() if rw_tryenter() fails.

Reply:
Good point. That signalling mechanism did not occur to me.